### PR TITLE
Start deprecation of --record flag

### PR DESCRIPTION
**Pull request by soltysh**
#### What type of PR is this?
/kind cleanup
/kind deprecation
/sig cli
/priority backlog

#### What this PR does / why we need it:
This is a follow up from a discussion at [SIG-CLI](https://docs.google.com/document/d/1r0YElcXt6G5mOWxwZiXgGu_X6he3F--wKwg-9UBc29I/edit#heading=h.2zzbdmokauz2) some time ago and email send to [k/dev](https://groups.google.com/g/kubernetes-dev/c/g8pm9bg1-Lo). I haven't heard any complaints and with [new mechanism which sends commands through HTTP headers](https://github.com/kubernetes/enhancements/issues/859) entering beta and very rich auditing functionality available in kubernetes it's time to deprecate `--record` flag which wasn't used consistently across all commands and is rather confusing to a lot of users. 

#### Special notes for your reviewer:
/assign @eddiezane 

#### Does this PR introduce a user-facing change?
```release-note
Deprecate --record flag in kubectl. The --record flag is being replaced with the mechanism from https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/859-kubectl-headers which annotates HTTP requests with kubectl command details.

```

### CODE REVIEW

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/record_flags.go [hunk 1]
```diff
@@ -93,6 +93,7 @@ func (f *RecordFlags) AddFlags(cmd *cobra.Command) {
 
 	if f.Record != nil {
 		cmd.Flags().BoolVar(f.Record, "record", *f.Record, "Record current kubectl command in the resource annotation. If set to false, do not record the command. If set to true, record the command. If not set, default to updating the existing annotation value only if one already exists.")
+		cmd.Flags().MarkDeprecated("record", "--record will be removed in the future")
```
Comment (sftim) [inline staging/src/k8s.io/cli-runtime/pkg/genericclioptions/record_flags.go:96]:
```suggestion
		cmd.Flags().MarkDeprecated("record", "--record will be removed in a future Kubernetes release")
```
?

**Comment by sftim** [comment]
Changelog nit; can I suggest:
```
Deprecate --record flag in kubectl. The --record flag is being replaced with the mechanism from https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/859-kubectl-headers which annotates HTTP requests with kubectl command details.
```

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/record_flags.go (hunk 1 above)
Comment (soltysh) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/record_flags.go:96]: That's consistent with all the rest of deprecation messages so I left is as is.

**Comment by soltysh** [comment]
> Changelog nit; can I suggest:

Good point - updated.

**Review by eddiezane — APPROVED** [review]

**Comment by eddiezane** [comment]
Flakes might be https://github.com/kubernetes/kubernetes/issues/100331 https://github.com/kubernetes/kubernetes/issues/97071.

**Comment by BenTheElder** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/40422

**Comment by abyssmemes** [comment]
Did you have alternative?

**Comment by BenTheElder** [comment]
see the section "What this PR does / why we need it:" in the PR description https://github.com/kubernetes/kubernetes/pull/102873#issue-921135374

TLDR: [audit logging](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/) + https://github.com/kubernetes/enhancements/issues/859

**Comment by tanvp112** [comment]
@BenTheElder how does the audit logging apply here, hopefully this doesn't mean we need to browse thru log for history command?

The other link sounds more reasonable, but any idea how to use it with `kubectl rollout history` and `undo`?

Hopefully `--record` will not be remove until it is clear how to use the above alternative. This can get even more confuse than `--record` alone.

**Comment by squaricdot** [comment]
Would like to mention that using the record function is also still part of the linuxfoundation CKA curriculum at the moment.

**Comment by vjunior1981** [comment]
@olwenya just tested like this and it worked:
```
~$ kubectl create deploy nginx --image=nginx --replicas=2
deployment.apps/nginx created
~$ kubectl set image deploy/nginx nginx=nginx:1.19
deployment.apps/nginx image updated
~$ kubectl annotate deploy/nginx kubernetes.io/change-cause='update image to 1.19'
deployment.apps/nginx annotated
~$ kubectl rollout history deploy/nginx
deployment.apps/nginx 
REVISION  CHANGE-CAUSE
1         <none>
2         update image to 1.19
```

Not perfect, but is an alternative.
